docs/blog/20240918/20240918.md

```diff
@@ -0,0 +1,71 @@
+
+## X-IO 
+X-IO is implemented using shared memory processing and unlocked producer/consumer rings between callers on the same nodes as the "X-IO stack". This approach is a much simpler alternative to kernel network stacking, enabling zero-copy packet transfers and providing a substantial improvement in data plane performance.
+### Unlocked Shared Memory Communication
+Two key elements are required to support shared memory communication
+- Shared memory buffer pool
+- Packet descriptor delivery mechanism
+The shared memory pool provides a shareable backend to store the payload for module access. Packet descriptor delivery delivers pointers to payloads in shared memory between different modules, rather than moving payloads.
+
+![An architectural overview of X-IO](./architecture.png)
+
+### Shared Memory Pools
+X-IO utilizes the X-IO Manager to manage the initialization of shared memory pools. These pools contain a specific number of shared memory buffers. The X-IO Manager runs as the DPDK master process, which grants it the privilege to create memory pools in the Linux file system.
+
+### Unlocked Packet Descriptor Transfer
+To efficiently transfer packet descriptors between caller' X-IO stacks, X-IO utilizes DPDK's RTE RING as a high-speed IPC channel. To minimize latency and overhead, X-IO implements a non-locked producer/consumer ring design. Each X-IO stack has a pair of RTE RINGs for receiving and transmitting, which are shared with the XIO manager. This eliminates the need for locks and allows the X-IO Manager to facilitate descriptor transfer between different X-IO stacks.
+
+### Raw I/O primitives in X-IO: zero-copy interfaces
+X-IO offers raw I/O primitives that enable zero-copy communication between caller. These primitives are implemented using DPDK's RTE RING and Mempool APIs.
+
+![communication](./communication.png)
+The image demonstrates zero-copy and shared memory communication in X-IO using the original I/O primitive. It involves the source module writing the payload to a memory buffer and adding the descriptor to the caller's TX ring. The X-IO manager retrieves these descriptors from the caller's TX ring, parses the routing information, and places the descriptors into the target module's RX ring. This design eliminates message duplication, reduces overhead, and improves efficiency compared to core-based networks.
+
+X-IO provides a high-performance interface for 5GC functions, utilizing lock-free shared memory processing. This improves the performance of the control plane, especially in achieving low latency.
+
+L25GC+ utilizes X-IO to enhance the support for 3GPP SBI. In comparison to the existing core-based SBI in free5GC, X-IO offers a superior and high-performance interface between the 5GC functionality and the underlying unlocked shared memory processing.
+
+Here are the improvements of L25GC+ with a commercial testbed:
```

Comment:
commercial testbed?

Reply:
This sentence appears in L25GC+ slides form [this](https://shixiongqi.github.io/_pages/slides/L25GC+_slides.pdf) in page11.

Reply:
@tim1207 
Please provide the details of commercial testbed in the blog:
![image](https://github.com/user-attachments/assets/80788321-c33b-4faa-af51-9b1a9043db7f)

Reply:
Got it!